### PR TITLE
Fix #6154: Make SolanaWeb3 interface not accessible by Pages

### LINKED ISSUE #6154: Enable Solana Dapps support by default
### Description: 

Enable Solana Dapps support by default

### PR DESCRIPTION
**Pull request by Brandon-T**
## Security Review
- https://github.com/brave/security/issues/1040

## Summary of Changes
- Hide SolanaWeb3 in anonymous namespace variable only known to the Swift code by name.

The way the code works is that it generates a random UUIDv4 (we call this the namespace).
Then we inject JS code into that namespace like:

```javascript
// Global variable with random namespace name.
// Pages cannot enumerate variables so it isn't possible to know that this variable exists
// unless the page bruteforces all UUIDv4's and guesses this variable name.
// A 1 in 2.71 quintillion probability.

let W02C00302CE624DAEAD70FDEE19594856;

__firefox__.execute(function($, $Object) {
    // Pages cannot access local scope variables or code!
    inject_our_SolanaWeb3_Javascript_Here_as_LOCAL_scope();

    // Initialize our namespace to a hardened dictionary
    W02C00302CE624DAEAD70FDEE19594856 = $({
        "solanaWeb3": $(solanaWeb3)  // Hardened solanaWeb3 local scope object
    });
});
```

Now whenever our code needs to make calls to `solanaWeb3`, we have to do:
```javascript
W02C00302CE624DAEAD70FDEE19594856.solanaWeb3.WHATEVER_HERE

// Example:
let pubKey = new W02C00302CE624DAEAD70FDEE19594856.solanaWeb3.PublicKey(...);

return pubKey;  // Maybe $(pubKey) to harden it? But not sure if objects should be mutable.
```

Any page attempting to do `new solanaWeb3.PublicKey` will find that it doesn't exist and can't be accessed since it is not in global scope. The page would have to know our UUID generated variable name.
We could proxy all calls to our `W02C00302CE624DAEAD70FDEE19594856.solanaWeb3` as well, so that all returned objects/functions/values are immutable. Just not sure that this is needed or WHY.

This pull request fixes #6154

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New or updated UI has been tested across:
  - [x] Light & dark mode
  - [x] Different size classes (iPhone, landscape, iPad)
  - [x] Different dynamic type sizes

## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

**Comment by thypon** [comment]
A bunch of questions here:

- Is this preventing the page from enumerating the objects defined on the `window|global` and finding this specific object?
- Is it possible to redefine that particular object?
- Do we incur in any significant performance penalty proxy wrapping everything in that object?

**Comment by stoletheminerals** [comment]
If someone wants to test dynamically don't forget to enable `isSolanaDappsEnabled` feature flag in `WalletConstants.swift`

File: Client/Frontend/Sync/WebFilter/Bookmarks/WalletHandler.swift [hunk 1]
```diff
@@ -0,0 +1,54 @@
+//
+//  File.swift
+//  
+//
+//  Created by Brandon on 2022-10-31.
+//
```
Comment (StephenHeaps) [inline Client/Frontend/Sync/WebFilter/Bookmarks/WalletHandler.swift]: Needs updated to use Brave copyright comment

File: Sources/BraveWallet/WalletConstants.swift [hunk 2]
```diff
@@ -97,6 +97,6 @@ struct WalletConstants {
 }
 
 public struct WalletDebugFlags {
-  public static let isSolanaDappsEnabled: Bool = false
+  public static let isSolanaDappsEnabled: Bool = true
```
Comment (StephenHeaps) [inline Sources/BraveWallet/WalletConstants.swift]: Fine until approved, but we should restore this to false before merging. We'll remove this flag and enable Solana Dapps by default in https://github.com/brave/brave-ios/issues/6154 with full Solana Dapps security review after this PR merges 🙂

**Review by StephenHeaps — COMMENTED** [review]
Functionality wise working great, thanks 👍. Unable to see `solanaWeb3` in Safari debugger . Couple small comments to address before merging but otherwise lgtm.

File: Client/Frontend/Sync/WebFilter/Bookmarks/WalletHandler.swift (hunk 1 above)
Comment (StephenHeaps) [reply Client/Frontend/Sync/WebFilter/Bookmarks/WalletHandler.swift]: Actually seems like this file might not be needed anymore / not in this PR?

File: Client/Frontend/Browser/UserScriptManager.swift [hunk 3]
```diff
@@ -239,12 +273,18 @@ class UserScriptManager {
           let solanaWeb3Script = solanaWeb3Script else {
       return
     }
-    // inject the Solana Web3 Library
-    await webView.evaluateSafeJavaScript(
-      functionName: solanaWeb3Script,
+    
+    let (result, error) = await webView.evaluateSafeJavaScript(
+      functionName: """
+      console.log(\(UserScriptManager.walletSolanaNameSpace));
```
Comment (stoletheminerals) [inline Client/Frontend/Browser/UserScriptManager.swift]: Do we need this?

**Comment by kdenhartog** [comment]
curious what the answers are to these questions: https://github.com/brave/brave-ios/pull/6264#issuecomment-1293649596

File: Client/Frontend/Browser/UserScriptManager.swift (hunk 3 above)
Comment (Brandon-T) [reply Client/Frontend/Browser/UserScriptManager.swift]: Not needed.

**Comment by Brandon-T** [comment]
> A bunch of questions here:
> 
> * Is this preventing the page from enumerating the objects defined on the `window|global` and finding this specific object?
> * Is it possible to redefine that particular object?
> * Do we incur in any significant performance penalty proxy wrapping everything in that object?

The page cannot enumerate the hidden solanaWeb3 object as it is local and they can't get it as it's wrapped in a UUID generated variable name which the page would take forever to guess.

The page cannot redefine that object because it doesn't know its name since it is randomly generated.

No performance penalty for wrapping it in that object.

ATM, the only thing the page can do is do something like: `Object.defineProperty(solana.publicKey, 'toBase58', {...}` but that is being addressed now and I'm freezing the objects returned to the page.

File: Client/Frontend/Sync/WebFilter/Bookmarks/WalletHandler.swift (hunk 1 above)
Comment (StephenHeaps) [reply Client/Frontend/Sync/WebFilter/Bookmarks/WalletHandler.swift]: This file is unused, should be removed if not needed

**Review by StephenHeaps — COMMENTED** [review]
Everything working as expected for me. Just need to remove the unused `WalletHandler.swift` file.

**Review by StephenHeaps — APPROVED** [review]

**Review by StephenHeaps — CHANGES_REQUESTED** [review]
Resolved in [0ff753b](https://github.com/brave/brave-ios/pull/6264/commits/0ff753b5a1880b1c8cfe42894bf14d808e12966c).

~Spoke too soon, _success_ flow for `Sign Transaction` & `Sign All Transactions (single)` are failing for me now; seems related to `solanaWeb3.Transaction` which both of `signTransaction` & `signAllTransactions` requests should return.~

1. Open pwgoom.csb.app
2. Connect and tap `Sign Transaction`
3. Tap `Continue`
4. Tap `Sign`

https://user-images.githubusercontent.com/5314553/203433063-5b423176-f6b2-4bf2-8fbf-6f3f7b0cd93f.mp4

**Review by StephenHeaps — APPROVED** [review]

**Review by nuo-xu — APPROVED** [review]
PR lgtm. Tested it with test dapp. all the requests look and being handled correctly.

**Comment by kdenhartog** [comment]
> The page cannot redefine that object because it doesn't know its name since it is randomly generated.

I think I've found a way to discover the value of the object via the console (note the `W5...` is the value of `UserScriptManager.walletSolanaNameSpace`), but it doesn't seem programatically accessible from my testing.
Screenshot 2022-11-29 at 7 49 45 PM

Additionally, I wasn't able to properly pollute the object either. So AFAICT this solution should be G2G. I'm going to approve. @thypon or @stoletheminerals feel free to close this one out if you're alright with the current solution as well.

**Review by kdenhartog — APPROVED** [review]

**Comment by stoletheminerals** [comment]
> I think I've found a way to discover the value of the object via the console (note the `W5...` is the value of `UserScriptManager.walletSolanaNameSpace`), but it doesn't seem programatically accessible from my testing.

Good catch @kdenhartog. It's possible to access it programatically by calling `solana.signMessage.toString()`. @Brandon-T we may need to overwrite it the same way you did for scripts with message handlers.

**Comment by Brandon-T** [comment]
> > I think I've found a way to discover the value of the object via the console (note the `W5...` is the value of `UserScriptManager.walletSolanaNameSpace`), but it doesn't seem programatically accessible from my testing.
> 
> Good catch @kdenhartog. It's possible to access it programatically by calling `solana.signMessage.toString()`. @Brandon-T we may need to overwrite it the same way you did for scripts with message handlers.

@stoletheminerals 
Fixed and pushed. It was a regression but I made the code much more robust and detection of what should be overridden is much better :)
Also I made the namespace inaccessible in stack-traces, so if an exception happens during injection or even as a result of a function call, the namespace name will not be in the stack-trace.

https://github.com/brave/brave-ios/pull/6264/commits/cac7c8aaf03a321fff0a87f7f8a4ae81d3cae9cb and https://github.com/brave/brave-ios/pull/6264/commits/d7675a7412a0ad055a96ddb8bfb7be580fd93b1d

Rebased.

**Comment by kdenhartog** [comment]
> > I think I've found a way to discover the value of the object via the console (note the `W5...` is the value of `UserScriptManager.walletSolanaNameSpace`), but it doesn't seem programatically accessible from my testing.
> 
> Good catch @kdenhartog. It's possible to access it programatically by calling `solana.signMessage.toString()`. @Brandon-T we may need to overwrite it the same way you did for scripts with message handlers.

Nice job finding a way to get it programatically accessible, I'm glad I posted it here then. Seems like @Brandon-T fix for it should address it.

Something else odd I found was that `window.solana.publicKey._bn.*` has quite a few math functions underneath it that were usable, but I couldn't do anything malicious with it. For example, `window.solana.publicKey._bn.nmul(0)` sets the `_bn` object to all 0s.

From a cryptography perspective that shouldn't have any affect since this is the public key and the EdDSA functions are designed so that they shouldn't be exploitable with malicious public keys. I do want to play around with it a bit more though because it may be possible to use this to bypass permissions for using keys or something like that. @thypon curious if you or @stoletheminerals can come up with a way to abuse this. To me, this just seems like a side effect of web3.js making their publicKey object extend a BN.js object and would actually be a vulnerability of their library if this is possible to abuse.

File: Client/Frontend/Browser/UserScriptManager.swift (hunk 3 above)
Comment (stoletheminerals) [reply Client/Frontend/Browser/UserScriptManager.swift]: @Brandon-T can you remove it then? 🙂

**Review by stoletheminerals — APPROVED** [review]
Just one minor logging issue. `SolanaWeb3` looks secure now. Though some `solana` properties are overwritable, I believe we don't want to freeze them for the same reason we don't freeze `ethereum` object.
Looks good, amazing job @Brandon-T

**Review by thypon — APPROVED** [review]

File: Client/Frontend/Browser/UserScriptManager.swift (hunk 3 above)
Comment (Brandon-T) [reply Client/Frontend/Browser/UserScriptManager.swift]: Removed @stoletheminerals

**Pull request merged**